File: src/main/java/io/github/rmuhamedgaliev/arcana/core/LocalizedText.java
package io.github.rmuhamedgaliev.arcana.core;

import java.util.HashMap;
import java.util.Map;

/**
 * Class for handling text in multiple languages.
 */
public class LocalizedText {
    private final Map<Language, String> texts;
    private Language defaultLanguage;

    public LocalizedText() {
        this.texts = new HashMap<>();
        this.defaultLanguage = Language.EN;
    }

    public LocalizedText(Map<Language, String> texts) {
        this.texts = new HashMap<>(texts);
        this.defaultLanguage = Language.EN;
    }

    public void setText(Language language, String text) {
        texts.put(language, text);
    }

    public String getText(Language language) {
        String text = texts.get(language);
        if (text == null) {
            // Fallback to default language if text is not available in requested language
            text = texts.get(defaultLanguage);
            if (text == null && !texts.isEmpty()) {
                // Fallback to any available language if text is not available in default language
                text = texts.values().iterator().next();
            }
        }
        return text != null ? text : "";
    }

    public void setDefaultLanguage(Language defaultLanguage) {
        this.defaultLanguage = defaultLanguage;
    }

    public Language getDefaultLanguage() {
        return defaultLanguage;
    }

    public boolean hasText(Language language) {
        return texts.containsKey(language);
    }

    public Map<Language, String> getAllTexts() {
        return new HashMap<>(texts);
    }
}

File: src/main/kotlin/io/github/rmuhamedgaliev/arcana/domain/model/LocalizedText.kt
package io.github.rmuhamedgaliev.arcana.domain.model

import com.fasterxml.jackson.annotation.JsonIgnore

/**
 * Class for storing text in multiple languages.
 * Supports variable substitution in text.
 */
class LocalizedText {
    private val texts: MutableMap<Language, String> = mutableMapOf()
    
    /**
     * Set text for a language.
     *
     * @param language The language
     * @param text The text
     */
    fun setText(language: Language, text: String) {
        texts[language] = text
    }
    
    /**
     * Get text for a language.
     *
     * @param language The language
     * @return The text, or null if not set
     */
    fun getText(language: Language): String? {
        return texts[language]
    }
    
    /**
     * Get text for a language with variable substitution.
     *
     * @param language The language
     * @param variables The variables to substitute
     * @return The text with variables substituted, or null if not set
     */
    fun getText(language: Language, variables: Map<String, Any>): String? {
        val text = texts[language] ?: return null
        
        return text.replace(Regex("\\{([^}]+)\\}")) { matchResult ->
            val variableName = matchResult.groupValues[1]
            variables[variableName]?.toString() ?: "{$variableName}"
        }
    }
    
    /**
     * Check if text is set for a language.
     *
     * @param language The language
     * @return True if text is set, false otherwise
     */
    fun hasText(language: Language): Boolean {
        return texts.containsKey(language)
    }
    
    /**
     * Get all texts.
     *
     * @return A map of all texts
     */
    @JsonIgnore
    fun getAllTexts(): Map<Language, String> {
        return texts.toMap()
    }
    
    /**
     * Get text for a language, or a fallback language if not set.
     *
     * @param language The language
     * @param fallbackLanguage The fallback language
     * @return The text, or null if not set in either language
     */
    fun getTextWithFallback(language: Language, fallbackLanguage: Language = Language.EN): String? {
        return texts[language] ?: texts[fallbackLanguage]
    }
    
    /**
     * Get text for a language with variable substitution, or a fallback language if not set.
     *
     * @param language The language
     * @param variables The variables to substitute
     * @param fallbackLanguage The fallback language
     * @return The text with variables substituted, or null if not set in either language
     */
    fun getTextWithFallback(language: Language, variables: Map<String, Any>, fallbackLanguage: Language = Language.EN): String? {
        val text = texts[language] ?: texts[fallbackLanguage] ?: return null
        
        return text.replace(Regex("\\{([^}]+)\\}")) { matchResult ->
            val variableName = matchResult.groupValues[1]
            variables[variableName]?.toString() ?: "{$variableName}"
        }
    }
}
